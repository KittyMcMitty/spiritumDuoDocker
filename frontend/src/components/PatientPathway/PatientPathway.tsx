import React, { useMemo } from 'react';
import { useQuery, gql } from '@apollo/client';
import { BarStackHorizontal } from '@visx/shape';
import { Text } from '@visx/text';
import { SeriesPoint } from '@visx/shape/lib/types';
import { Group } from '@visx/group';
import { AxisTop } from '@visx/axis';
import { scaleBand, scaleLinear, scaleOrdinal } from '@visx/scale';
import { withTooltip, Tooltip, defaultStyles } from '@visx/tooltip';
import { WithTooltipProvidedProps } from '@visx/tooltip/lib/enhancers/withTooltip';
import { ParentSize } from '@visx/responsive';

import { getPatientWithReferrals } from './__generated__/getPatientWithReferrals';
import { MilestoneState } from '../../__generated__/globalTypes';

const chartBlue1 = '#003087';
const red = '#FF2D00';
const chartBlue2 = '#99C7EB';
const NHSBlue = '#005EB8';
const chartGrey1 = '#919EA8';
const chartGrey2 = '#DDE1E4';
const purple3 = '#a44afe';
const background = '#F1F4F5';

const barCornerCurve = '0.7%';
const friendlyNames = {
  request: 'Awaiting Result',
  result: 'Awaiting Decision',
  empty: 'Stalled',
};

const tooltipStyles = {
  ...defaultStyles,
  minWidth: 60,
  backgroundColor: 'rgba(0,0,0,0.9)',
  color: 'white',
  zIndex: 4,
};

const defaultMargin = { top: 20, left: 20, right: 20, bottom: 20 };

// Helper functions
const getPatientName = (d: PatientData): string => d.name;
const getDatePortionString = (date: Date): string => date.toISOString().split('T', 1)[0];
const msToDays = (ms: number): number => ms / 86400000;

let tooltipTimeout: number;

export const GET_PATIENT_WITH_REFERRALS_QUERY = gql`
query getPatientWithReferrals($hospitalNumber: String!) {
  getPatient(hospitalNumber: $hospitalNumber) {
    id
    firstName
    lastName
    dateOfBirth
    hospitalNumber
    onPathways { 
      id
      referredAt
      milestones {
          id
          milestoneType {
            id
            name
          }
          addedAt
          updatedAt
          currentState
          forwardDecisionPoint {
            id
            addedAt
          }
      }
    }
  }
}`;

interface Period {
  duration: number;
}

type RequestPeriod = {
  type: 'request';
  requests: string[];
} & Period;

type ResultPeriod = {
  type: 'result';
  requests: string[];
  results: string[];
} & Period;

type EmptyPeriod = {
  type: 'empty';
} & Period;

type PatientPeriod = EmptyPeriod | RequestPeriod | ResultPeriod;

interface PatientData {
  name: string;
  periods: {
    [index: string]: PatientPeriod;
  };
}

type TooltipData = {
  bar: SeriesPoint<PatientData>;
  key: string;
  index: number;
  height: number;
  width: number;
  x: number;
  y: number;
  color: string;
};

type BarStackHorizontalProps = {
  data: PatientData[];
  maxDays: number;
  showName: boolean;
  width: number;
  height: number;
  margin?: { top: number; right: number; bottom: number; left: number };
};

const PathwayVisualisation = withTooltip<BarStackHorizontalProps, TooltipData>(
  ({
    data,
    maxDays,
    showName,
    width,
    height,
    margin = defaultMargin,
    tooltipOpen,
    tooltipLeft,
    tooltipTop,
    tooltipData,
    hideTooltip,
    showTooltip,
  }: BarStackHorizontalProps & WithTooltipProvidedProps<TooltipData>) => {
    // memoise all this to prevent scale recalculation on re-render
    // bounds
    const xMax = useMemo(
      () => width - margin.left - margin.right,
      [margin.left, margin.right, width],
    );
    // Proportion of x axis to use for scale / leave for right pill
    const scaleXMax = useMemo(() => (xMax > 500 ? xMax * 0.80 : 0), [xMax]);
    // magic number to stay in proportion
    const yMax = useMemo(() => (scaleXMax !== 0 ? xMax / 10 : 100), [xMax, scaleXMax]);

    const rightPillX = scaleXMax + (scaleXMax * 0.01);
    const rightPillWidth = scaleXMax !== 0
      ? xMax * 0.2
      : 200;

    const rightTextX = scaleXMax !== 0
      ? xMax * 0.8
      : 100;
    const rightTextXOffset = scaleXMax !== 0
      ? rightPillWidth / 2
      : 0;
    const rightTextWidth = scaleXMax !== 0
      ? xMax * 0.2
      : 100;

    const tickValues = [0, maxDays / 4, maxDays / 2, maxDays * 0.75, maxDays];

    const dayScale = useMemo(() => {
      const dayScaleMax = maxDays || Math.max(
        ...data.map((d) => Object.keys(d.periods).reduce(
          (previous, current) => previous + d.periods[current].duration, 0,
        ), 0),
      );
      const scale = scaleLinear<number>({
        domain: [0, dayScaleMax],
        clamp: true,
      });
      scale.rangeRound([0, scaleXMax]);
      return scale;
    }, [data, maxDays, scaleXMax]);

    const patientScale = useMemo(() => {
      const scale = scaleBand<string>({
        domain: data.map(getPatientName),
        paddingOuter: 0.2,
        paddingInner: 0.66,
      });
      scale.rangeRound([0, yMax]);
      return scale;
    }, [data, yMax]);

    const patientColourScale = useMemo(() => scaleOrdinal({
      domain: ['request', 'result', 'empty'],
      range: [chartBlue2, NHSBlue, red],
    }), []);

    const patientWithMostPeriods = useMemo(() => data.reduce((previous, current) => (
      Object.keys(previous.periods).length > Object.keys(current.periods).length
        ? previous
        : current
    )), [data]);
    const pKeys = Object.keys(patientWithMostPeriods.periods);

    return width < 10 ? null : (
      <div>
        <svg width={ width } height={ height }>
          <rect width={ width } height={ height } fill="white" />
          <Group top={ margin.top } left={ margin.left }>
            <BarStackHorizontal<PatientData, string>
              value={ (d, k) => d.periods[k]?.duration }
              data={ data }
              keys={ pKeys }
              height={ yMax }
              y={ getPatientName }
              xScale={ dayScale }
              yScale={ patientScale }
              color={ (key, index) => {
                const type = data[index].periods[key]?.type;
                const colour = type
                  ? patientColourScale(type)
                  : undefined;
                return colour || background;
              } }
            >
              {(barStacks) => (
                <React.Fragment key={ `barstack-horizontal-${barStacks}` }>
                  {
                    barStacks[0].bars.map((b) => {
                      const lastPeriodKey = Object.keys(b.bar.data.periods)
                        .sort((first, second) => parseInt(second, 10) - parseInt(first, 10))[0];
                      const lastPeriod = b.bar.data.periods[lastPeriodKey];
                      let pillText: string;
                      switch (lastPeriod.type) {
                        case 'request': {
                          const p = lastPeriod as RequestPeriod;
                          const plusText = p.requests.length > 1
                            ? `+${p.requests.length - 1}`
                            : '';
                          pillText = `Awaiting ${p.requests[0]} ${plusText}`;
                          break;
                        }
                        case 'result': {
                          const p = lastPeriod as ResultPeriod;
                          const plusText = p.results.length > 1
                            ? `+${p.results.length - 1}`
                            : '';
                          pillText = `Awaiting ${p.results[0]} Acknowledgement ${plusText}`;
                          break;
                        }
                        default:
                          pillText = 'Stalled';
                      }
                      return (
                        <Group key={ `${b.bar.data.name}-${String(Symbol(b.bar.data.name))}` }>
                          {/* Row Background */}
                          <rect
                            width={ width }
                            height={ b.height * 2.5 }
                            x={ b.x - margin.left }
                            y={ b.y - b.height }
                            fill="white"
                          />
                          {
                            /* Patient name */
                            showName
                              ? (
                                <Text y={ b.y } dy={ -5 } fontSize="1.5rem" fontWeight={ 750 } scaleToFit>
                                  {b.bar.data.name}
                                </Text>
                              )
                              : undefined
                          }
                          {/* Bars background */}
                          <rect
                            width={ scaleXMax }
                            height={ b.height }
                            y={ b.y }
                            fill="#D8DDE0"
                            rx={ barCornerCurve }
                          />
                          {/* Right pill */}
                          <rect
                            height={ b.height * 1.5 }
                            y={ b.y - (b.height / 2) }
                            rx={ barCornerCurve }
                            x={ rightPillX }
                            fill={ patientColourScale(lastPeriod.type) }
                            width={ rightPillWidth }
                          />
                          {/* Right pill text */}
                          <Text
                            fill="white"
                            textAnchor="middle"
                            fontSize="0.75rem"
                            y={ b.y }
                            x={ rightTextX }
                            dy={ b.height / 1.5 }
                            dx={ rightTextXOffset }
                            width={ rightTextWidth }
                          >
                            {pillText}
                          </Text>
                        </Group>
                      );
                    })
                  }
                  {
                  barStacks.map((barStack) => barStack.bars.map((bar) => (
                    <rect
                      key={ `barstack-horizontal-${barStack.index}-${bar.index}` }
                      rx={ barCornerCurve }
                      x={ bar.x }
                      y={ bar.y }
                      width={ bar.width }
                      height={ bar.height }
                      fill={ bar.color }
                      onMouseLeave={ () => {
                        tooltipTimeout = window.setTimeout(() => {
                          hideTooltip();
                        }, 300);
                      } }
                      onMouseMove={ () => {
                        if (tooltipTimeout) clearTimeout(tooltipTimeout);
                        const top = bar.y + margin.top;
                        const left = bar.x + bar.width + margin.left;
                        showTooltip({
                          tooltipData: bar,
                          tooltipTop: top,
                          tooltipLeft: left,
                        });
                      } }
                    />
                  )))
                }
                </React.Fragment>
              )}
            </BarStackHorizontal>
            {
              scaleXMax !== 0
                ? (
                  <AxisTop
                    numTicks={ 4 }
                    tickValues={ tickValues }
                    hideAxisLine
                    top={ margin.top - 2 }
                    scale={ dayScale }
                    stroke={ purple3 }
                    tickStroke={ purple3 }
                    tickLabelProps={ () => ({
                      fill: purple3,
                      fontSize: 11,
                      textAnchor: 'middle',
                    }) }
                  />
                )
                : undefined
              }
          </Group>
        </svg>
        {tooltipOpen && tooltipData && (
          <Tooltip top={ tooltipTop } left={ tooltipLeft } style={ tooltipStyles }>
            <div>
              <h2>{friendlyNames[tooltipData.bar.data.periods[tooltipData.key].type]}</h2>
            </div>
            <div>{
              (
                tooltipData.bar.data.periods[tooltipData.key] as RequestPeriod
              ).requests !== undefined
                ? (
                  <>
                    <h5>Requests</h5>
                    {(
                      tooltipData.bar.data.periods[tooltipData.key] as RequestPeriod
                    ).requests.map((r) => <>{r}<br /></> )}
                  </>
                )
                : false
            }
            </div>
            <div>{
              (tooltipData.bar.data.periods[tooltipData.key] as ResultPeriod).results !== undefined
                ? (
                  <>
                    <h5>Results</h5>
                    {(
                      tooltipData.bar.data.periods[tooltipData.key] as ResultPeriod
                    ).results.map((r) => <>{r}<br /></> )}
                  </>
                )
                : false
            }
            </div>
          </Tooltip>
        )}
      </div>
    );
  },
);

export interface PatientPathwayProps {
  hospitalNumber: string;
  showName?: boolean;
  maxDays?: number;
}

const PatientPathway = (
  { hospitalNumber, maxDays = 70, showName }: PatientPathwayProps,
): JSX.Element => {
  const {
    loading, error, data,
  } = useQuery<getPatientWithReferrals>(
    GET_PATIENT_WITH_REFERRALS_QUERY,
    { variables: { hospitalNumber } },
  );

  const patientData: PatientData | undefined = useMemo(() => {
    if (data) {
      interface RequestEvent {id: string; name: string}
      interface EventDates {
        [date: string]: {
          requests?: Set<RequestEvent>;
          results?: Set<RequestEvent>;
          acknowledgements?: Set<RequestEvent>;
        }
      }
      const events: EventDates = {};
      data.getPatient?.onPathways?.[0].milestones?.forEach((ms) => {
        let startDay = events[getDatePortionString(ms.addedAt)];
        if (!startDay) {
          startDay = {};
          events[getDatePortionString(ms.addedAt)] = startDay;
        }
        const event = {
          id: ms.id,
          name: ms.milestoneType.name,
        };
        startDay.requests !== undefined
          ? startDay.requests.add(event)
          : startDay.requests = new Set([event]);

        if (ms.currentState === MilestoneState.COMPLETED) {
          let endDay = events[getDatePortionString(ms.updatedAt)];
          if (!endDay) {
            endDay = {};
            events[getDatePortionString(ms.updatedAt)] = endDay;
          }
          endDay.results !== undefined
            ? endDay.results.add(event)
            : endDay.results = new Set([event]);

          if (ms.forwardDecisionPoint) {
            let acknowledgedDay = events[getDatePortionString(ms.forwardDecisionPoint.addedAt)];
            if (!acknowledgedDay) {
              acknowledgedDay = {};
              events[getDatePortionString(ms.forwardDecisionPoint.addedAt)] = acknowledgedDay;
            }
            acknowledgedDay.acknowledgements !== undefined
              ? acknowledgedDay.acknowledgements.add(event)
              : acknowledgedDay.acknowledgements = new Set([event]);
          }
        }
      });

      const returnData: PatientData = {
        name: `${data.getPatient?.firstName} ${data.getPatient?.lastName}`,
        periods: {},
      };

      const sortedDates = Object.keys(events).sort();
      let periodKey = 0;
      if (sortedDates[0] !== getDatePortionString(data.getPatient?.onPathways?.[0].referredAt)) {
        const currentDate = new Date(data.getPatient?.onPathways?.[0].referredAt);
        const nextDate = new Date(sortedDates[0]);
        const timeDelta = Math.abs(currentDate.getTime() - nextDate.getTime());
        const emptyPeriod: EmptyPeriod = {
          type: 'empty',
          duration: msToDays(timeDelta),
        };
        returnData.periods[periodKey.toString()] = emptyPeriod;
        periodKey += 1;
      }

      const currentEvents: {
        requests: Set<RequestEvent>;
        results: Set<RequestEvent>;
      } = { requests: new Set(), results: new Set() };

      sortedDates.forEach((sd, index, array) => {
        const currentEventDate = events[sd];
        const currentDateMs = new Date(sd).getTime();
        const nextDateMs = index < array.length - 1
          ? new Date(array[index + 1]).getTime()
          : Date.now();
        const duration = msToDays(nextDateMs - currentDateMs);

        // Add new requests
        currentEventDate.requests?.forEach((request) => currentEvents.requests.add(request));
        // Add new results & delete from requests
        currentEventDate.results?.forEach((result) => {
          currentEvents.requests.delete(result);
          currentEvents.results.add(result);
        });
        // Remove acknowledgements from results
        currentEventDate.acknowledgements?.forEach((ack) => currentEvents.results.delete(ack));
        const type = (currentEvents.requests.size === 0 && currentEvents.results.size === 0)
          ? 'empty'
          : (currentEvents.requests.size > 0 && currentEvents.results.size === 0)
            ? 'request'
            : 'result';

        let period: PatientPeriod;
        switch (type) {
          case 'result':
            period = {
              type: 'result',
              duration: duration,
              results: Array.from(currentEvents.results).map((r) => r.name),
              requests: Array.from(currentEvents.requests).map((r) => r.name),
            };
            break;

          case 'request':
            period = {
              type: 'request',
              duration: duration,
              requests: Array.from(currentEvents.requests).map((r) => r.name),
            };
            break;

          default:
            period = {
              type: 'empty',
              duration: duration,
            };
        }
        returnData.periods[periodKey] = period;
        periodKey += 1;
      });
      return returnData;
    }
    return undefined;
  }, [data]);

  if (patientData) {
    return (
      <ParentSize debounceTime={ 0 }>
        {({ width: parentWidth, height: parentHeight }) => (
          <PathwayVisualisation
            showName={ showName || false }
            maxDays={ maxDays }
            data={ [patientData] }
            width={ parentWidth }
            height={ parentHeight }
          />
        )}
      </ParentSize>
    );
  }
  return <h1>Loading!</h1>;
};

export default PatientPathway;
