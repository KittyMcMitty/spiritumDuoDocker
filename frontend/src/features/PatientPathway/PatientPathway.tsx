import React, { useMemo } from 'react';
import PathwayVisualisation, { PatientVisData, patientToVisData, PATIENT_PARTS_FOR_GRP } from 'components/PathwayVisualisation';
import { gql, useQuery } from '@apollo/client';
import { ParentSize } from '@visx/responsive';

import { getPatientWithReferrals } from './__generated__/getPatientWithReferrals';

export const GET_PATIENT_WITH_REFERRALS_QUERY = gql`
  ${PATIENT_PARTS_FOR_GRP}
  query getPatientWithReferrals($hospitalNumber: String!) {
    getPatient(hospitalNumber: $hospitalNumber) {
      ...GrpPatientFields
    }
  }
`;

export interface PatientPathwayProps {
  hospitalNumber: string;
  showName?: boolean;
  showNationalNumber?: boolean;
  maxDays?: number;
}

const PatientPathway = (
  { hospitalNumber, maxDays = 70, showName, showNationalNumber }: PatientPathwayProps,
): JSX.Element => {
  const {
    loading, error, data,
  } = useQuery<getPatientWithReferrals>(
    GET_PATIENT_WITH_REFERRALS_QUERY,
    { variables: { hospitalNumber } },
  );

  const patientData: PatientVisData | undefined = useMemo(() => {
    if (data?.getPatient) {
      return patientToVisData(data.getPatient);
    }
    return undefined;
  }, [data]);

  if (patientData) {
    return (
      <ParentSize debounceTime={ 0 }>
        {({ width: parentWidth }) => (
          <PathwayVisualisation
            showName={ showName }
            axisBottom
            showNationalNumber={ showNationalNumber }
            maxDays={ maxDays }
            backgroundColor="white"
            data={ [patientData] }
            width={ parentWidth }
            margin={ { top: 0, right: 10, bottom: 0, left: 10 } }
          />
        )}
      </ParentSize>
    );
  }
  return <h1>Loading!</h1>;
};

export default PatientPathway;
