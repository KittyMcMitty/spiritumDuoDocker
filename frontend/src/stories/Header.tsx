import React from 'react';

import { Button } from './Button';
import './header.css';
import { PathwaySelector } from './PathwaySelector';

interface HeaderProps {
	user?: {};
}

export const Header = ({user}: HeaderProps) => (
	<nav className="navbar navbar-expand-lg navbar-light bg-light">
	<div className="container-fluid">
		<form className="d-flex">
			<input className="form-control me-2" type="search" placeholder="Hospital number" aria-label="Hospital number"/>
		</form>
		<div className="collapse navbar-collapse" id="navbarSupportedContent">
			<ul className="navbar-nav mb-2">
				<li className="nav-item">
					<a className="nav-link active" aria-current="page" href="#">Home</a>
				</li>
				<li className="nav-item">
					<a className="nav-link" href="#">Triage</a>
				</li>
				<li className="nav-item">
					<a className="nav-link" href="#">Clinic</a>
				</li>
				<li className="nav-item">
					<a className="nav-link" href="#">MDT</a>
				</li>
				<li className="nav-item">
					<a className="nav-link" href="#">Refer</a>
				</li>
			</ul>
		</div>
		<PathwaySelector options={["Lung cancer", "Bronchieactasis"]} onItemSelect={(data:string)=>console.log(data)}/>
	</div>
</nav>
);