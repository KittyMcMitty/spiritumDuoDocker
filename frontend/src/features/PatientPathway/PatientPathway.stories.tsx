/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import PatientPathway, { GET_PATIENT_WITH_REFERRALS_QUERY } from 'features/PatientPathway/PatientPathway';
import { NewMockSdApolloProvider } from 'test/mocks/mockApolloProvider';
import MockDate from 'test/mocks/mockDate';

import { DocumentNode } from 'graphql';
import { RequestHandler } from 'mock-apollo-client';
import { getPatientWithReferrals } from './__generated__/getPatientWithReferrals';
import { ClinicalRequestState } from '../../__generated__/globalTypes';

const apolloMock: {
  query: DocumentNode,
  mockFn: RequestHandler<getPatientWithReferrals, {hospitalNumber: string}>
}[] = [{
  query: GET_PATIENT_WITH_REFERRALS_QUERY,
  mockFn: () => Promise.resolve({
    data: {
      getPatient: {
        __typename: 'Patient',
        id: '1',
        firstName: 'John',
        lastName: 'Doe',
        dateOfBirth: new Date('1970-10-10'),
        hospitalNumber: 'fMRN1234567',
        nationalNumber: 'fNHS1234567890',
        onPathways: [{
          __typename: 'OnPathway',
          id: '1',
          referredAt: new Date('2021-08-14'),
          clinicalRequests: [
            {
              __typename: 'ClinicalRequest',
              id: '1',
              clinicalRequestType: { __typename: 'ClinicalRequestType', id: '1', name: 'CT-Thorax' },
              addedAt: new Date('2021-08-14'),
              updatedAt: new Date('2021-08-16'),
              currentState: ClinicalRequestState.COMPLETED,
              forwardDecisionPoint: {
                __typename: 'DecisionPoint',
                id: '1',
                addedAt: new Date('2021-08-24'),
              },
            },
            {
              __typename: 'ClinicalRequest',
              id: '2',
              clinicalRequestType: { __typename: 'ClinicalRequestType', id: '2', name: 'X-Ray' },
              addedAt: new Date('2021-08-14'),
              updatedAt: new Date('2021-08-20'),
              currentState: ClinicalRequestState.COMPLETED,
              forwardDecisionPoint: {
                __typename: 'DecisionPoint',
                id: '1',
                addedAt: new Date('2021-08-24'),
              },
            },
            {
              __typename: 'ClinicalRequest',
              id: '3',
              clinicalRequestType: { __typename: 'ClinicalRequestType', id: '3', name: 'Lung Function' },
              addedAt: new Date('2021-08-14'),
              updatedAt: new Date('2021-08-15'),
              currentState: ClinicalRequestState.COMPLETED,
              forwardDecisionPoint: {
                __typename: 'DecisionPoint',
                id: '1',
                addedAt: new Date('2021-08-24'),
              },
            },
            {
              __typename: 'ClinicalRequest',
              id: '4',
              clinicalRequestType: { __typename: 'ClinicalRequestType', id: '4', name: 'CT-Thorax' },
              addedAt: new Date('2021-08-24'),
              updatedAt: new Date('2021-08-30'),
              currentState: ClinicalRequestState.COMPLETED,
              forwardDecisionPoint: {
                __typename: 'DecisionPoint',
                id: '2',
                addedAt: new Date('2021-09-06'),
              },
            },
            {
              __typename: 'ClinicalRequest',
              id: '5',
              clinicalRequestType: { __typename: 'ClinicalRequestType', id: '5', name: 'MRI-Head' },
              addedAt: new Date('2021-08-24'),
              updatedAt: new Date('2021-09-02'),
              currentState: ClinicalRequestState.COMPLETED,
              forwardDecisionPoint: {
                __typename: 'DecisionPoint',
                id: '2',
                addedAt: new Date('2021-09-06'),
              },
            },
            {
              __typename: 'ClinicalRequest',
              id: '6',
              clinicalRequestType: { __typename: 'ClinicalRequestType', id: '6', name: 'EBUS' },
              addedAt: new Date('2021-08-24'),
              updatedAt: new Date('2021-09-04'),
              currentState: ClinicalRequestState.COMPLETED,
              forwardDecisionPoint: {
                __typename: 'DecisionPoint',
                id: '2',
                addedAt: new Date('2021-09-06'),
              },
            },
            {
              __typename: 'ClinicalRequest',
              id: '7',
              clinicalRequestType: { __typename: 'ClinicalRequestType', id: '7', name: 'CT-Thorax' },
              addedAt: new Date('2021-09-10'),
              updatedAt: new Date('2021-09-10'),
              currentState: ClinicalRequestState.WAITING,
              forwardDecisionPoint: null,
            },
            {
              __typename: 'ClinicalRequest',
              id: '8',
              clinicalRequestType: { __typename: 'ClinicalRequestType', id: '8', name: 'Bloods' },
              addedAt: new Date('2021-09-10'),
              updatedAt: new Date('2021-09-10'),
              currentState: ClinicalRequestState.WAITING,
              forwardDecisionPoint: null,
            },
          ],
        }],
      },
    },
  }),
}];

export default {
  title: 'Features/PatientPathway',
  component: PatientPathway,
  decorators: [(Story) => (
    <MockDate date="2021-09-20">
      <NewMockSdApolloProvider mocks={ apolloMock }>
        <div style={ { width: '95vw', height: '500px' } }>
          <Story />
        </div>
      </NewMockSdApolloProvider>
    </MockDate>
  )],
} as ComponentMeta<typeof PatientPathway>;

const Template: ComponentStory<typeof PatientPathway> = (args) => <PatientPathway { ...args } />;

export const Default = Template.bind({});
Default.args = {
  hospitalNumber: 'fMRN1234567',
  showName: true,
};

Default.parameters = {
  mocks: apolloMock,
};

export const WithoutName = Template.bind({});
WithoutName.args = {
  hospitalNumber: 'fMRN1234567',
  showName: false,
};
