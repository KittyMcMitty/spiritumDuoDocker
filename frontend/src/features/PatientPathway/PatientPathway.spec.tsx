import React from 'react';
import { composeStories } from '@storybook/testing-react';
import { render, waitFor } from '@testing-library/react';
import * as stories from './PatientPathway.stories';

const { Default } = composeStories(stories);

it('It should render Default correctly', async () => {
  const { container, getByRole } = render(<Default />);
  await waitFor(() => expect(getByRole('table', { name: /pathway\s+table/i })).toBeInTheDocument());
  expect(container).toMatchSnapshot();
});
