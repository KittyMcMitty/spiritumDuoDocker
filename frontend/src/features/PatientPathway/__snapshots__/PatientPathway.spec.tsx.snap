// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render Default correctly 1`] = `
<div>
  <div
    style="width: 95vw; height: 500px;"
  >
    <div
      style="width: 100%; height: 100%;"
    >
      <div
        style="position: relative; width: inherit; height: inherit;"
      >
        <div>
          <div
            class="visually-hidden"
            id="pathway-visualisation-desc"
          >
            Patient Pathway Visualisation
          </div>
          <svg
            aria-describedby="pathway-visualisation-desc"
            aria-details="accessible-table"
            height="235"
            width="280"
          >
            <rect
              fill="white"
              height="235"
              width="280"
            />
            <g
              class="visx-group"
              transform="translate(15, 35)"
            >
              <g
                class="visx-group"
                transform="translate(0, 0)"
              >
                <rect
                  fill="white"
                  height="170"
                  width="280"
                  x="-10"
                  y="-2"
                />
                <svg
                  font-size="1.5rem"
                  style="overflow: visible;"
                  x="0"
                  y="-5"
                >
                  <text
                    font-size="1.5rem"
                    font-weight="650"
                    text-anchor="start"
                    transform=""
                    y="66"
                  >
                    <tspan
                      dy="0em"
                      x="0"
                    >
                      John Doe, fMRN1234567, 10/10/1970
                    </tspan>
                  </text>
                </svg>
                <rect
                  fill="#D8DDE0"
                  height="68"
                  rx="0.7%"
                  width="0"
                  y="66"
                />
                <rect
                  fill="#99C7EB"
                  height="68"
                  rx="0.7%"
                  width="126"
                  x="0"
                  y="86"
                />
                <svg
                  font-size="1rem"
                  style="overflow: visible;"
                  x="0"
                  y="65.33333333333334"
                >
                  <text
                    fill="white"
                    font-size="1rem"
                    text-anchor="middle"
                    transform=""
                    x="63"
                    y="66"
                  >
                    <tspan
                      dy="0em"
                      x="63"
                    >
                      Awaiting CT-Thorax +1
                    </tspan>
                  </text>
                </svg>
              </g>
              <rect
                fill="#99C7EB"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#005EB8"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#005EB8"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#005EB8"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#99C7EB"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#005EB8"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#005EB8"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#005EB8"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#FF2D00"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
              <rect
                fill="#99C7EB"
                height="68"
                rx="0.7%"
                width="0"
                x="0"
                y="66"
              />
            </g>
          </svg>
          <div
            class="visually-hidden"
            id="accessible-table"
          >
            <table
              aria-label="Pathway Table"
            >
              <thead>
                <tr>
                  <th
                    id="tname"
                    rowspan="2"
                    scope="col"
                  >
                    Name
                  </th>
                  <th
                    id="thosp"
                    rowspan="2"
                    scope="col"
                  >
                    Hospital Number
                  </th>
                  <th
                    id="tdob"
                    rowspan="2"
                    scope="col"
                  >
                    Date Of Birth
                  </th>
                  <th
                    colspan="4"
                    scope="colgroup"
                    style="text-align: center;"
                  >
                    Periods
                  </th>
                </tr>
                <tr>
                  <th
                    id="pt"
                    scope="colgroup"
                  >
                    Period Type
                  </th>
                  <th
                    id="pt"
                    scope="colgroup"
                  >
                    Duration
                  </th>
                  <th
                    id="preq"
                    scope="colgroup"
                  >
                    Outstanding Requests
                  </th>
                  <th
                    id="pres"
                    scope="colgroup"
                  >
                    Awaiting Results
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <th
                    headers="tname"
                    id="rowname-1"
                    role="rowheader"
                    rowspan="10"
                    scope="rowgroup"
                  >
                    <button
                      role="link"
                      tabindex="0"
                      type="button"
                    >
                      John Doe
                    </button>
                  </th>
                  <td
                    headers="thosp rowname-1"
                  >
                    fMRN1234567
                  </td>
                  <td
                    headers="tdob rowname-1"
                  >
                    10/10/1970
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Result
                  </td>
                  <td>
                    CT-Thorax
                     
                    <br />
                    X-Ray
                     
                    <br />
                    Lung Function
                     
                    <br />
                  </td>
                  <td>
                    -
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Decision
                  </td>
                  <td>
                    CT-Thorax
                     
                    <br />
                    X-Ray
                     
                    <br />
                  </td>
                  <td>
                    Lung Function
                     
                    <br />
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Decision
                  </td>
                  <td>
                    X-Ray
                     
                    <br />
                  </td>
                  <td>
                    Lung Function
                     
                    <br />
                    CT-Thorax
                     
                    <br />
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Decision
                  </td>
                  <td />
                  <td>
                    Lung Function
                     
                    <br />
                    CT-Thorax
                     
                    <br />
                    X-Ray
                     
                    <br />
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Result
                  </td>
                  <td>
                    CT-Thorax
                     
                    <br />
                    MRI-Head
                     
                    <br />
                    EBUS
                     
                    <br />
                  </td>
                  <td>
                    -
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Decision
                  </td>
                  <td>
                    MRI-Head
                     
                    <br />
                    EBUS
                     
                    <br />
                  </td>
                  <td>
                    CT-Thorax
                     
                    <br />
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Decision
                  </td>
                  <td>
                    EBUS
                     
                    <br />
                  </td>
                  <td>
                    CT-Thorax
                     
                    <br />
                    MRI-Head
                     
                    <br />
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Decision
                  </td>
                  <td />
                  <td>
                    CT-Thorax
                     
                    <br />
                    MRI-Head
                     
                    <br />
                    EBUS
                     
                    <br />
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Stalled
                  </td>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                </tr>
                <tr>
                  <td>
                    -
                  </td>
                  <td>
                    -
                  </td>
                  <td
                    headers="pt"
                  >
                    Awaiting Result
                  </td>
                  <td>
                    CT-Thorax
                     
                    <br />
                    Bloods
                     
                    <br />
                  </td>
                  <td>
                    -
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
